solr/core/src/test/org/apache/solr/cloud/overseer/ZkStateReaderTest.java

```diff
@@ -196,4 +209,251 @@ public void testWatchedCollectionCreation() throws Exception {
+            new SolrNamedThreadFactory("zkStateReaderTest"));
+
+    try {
+      server.run();
+
+      zkClient = new SolrZkClient(server.getZkAddress(), OverseerTest.DEFAULT_CONNECTION_TIMEOUT);
+      ZkController.createClusterZkNodes(zkClient);
+
+      reader = new ZkStateReader(zkClient);
+      final ZkStateReader readerRef = reader;
+      reader.createClusterStateWatchersAndUpdate();
+      zkClient.makePath(ZkStateReader.COLLECTIONS_ZKNODE + "/c1", true);
+
+      // start another thread to constantly updating the state
+      final AtomicBoolean stopMutatingThread = new AtomicBoolean(false);
+      final ZkStateWriter writer = new ZkStateWriter(reader, new Stats());
+      final AtomicInteger updateCounts = new AtomicInteger(0);
+      final AtomicReference<Exception> updateException = new AtomicReference<>();
+      executorService.submit(
+          () -> {
+            try {
+              ClusterState clusterState = readerRef.getClusterState();
+              while (!stopMutatingThread.get()) {
+                DocCollection collection = clusterState.getCollectionOrNull("c1");
+                int currentVersion = collection != null ? collection.getZNodeVersion() : 0;
+                // create new collection
+                DocCollection state =
+                    new DocCollection(
+                        "c1",
+                        new HashMap<>(),
+                        Map.of(
+                            ZkStateReader.CONFIGNAME_PROP,
+                            ConfigSetsHandler.DEFAULT_CONFIGSET_NAME),
+                        DocRouter.DEFAULT,
+                        currentVersion);
+                ZkWriteCommand wc = new ZkWriteCommand("c1", state);
+                writer.enqueueUpdate(clusterState, Collections.singletonList(wc), null);
+                clusterState = writer.writePendingUpdates();
+              }
+            } catch (Exception e) {
+              updateException.set(e);
+            }
+            return null;
+          });
+      executorService.shutdown();
+
+      for (int i = 0; i < RUN_COUNT; i++) {
+        final CountDownLatch latch = new CountDownLatch(2);
+
+        // remove itself on 2nd trigger
+        DocCollectionWatcher dummyWatcher =
+            collection -> {
+              latch.countDown();
+              return latch.getCount() == 0;
+            };
+        reader.registerDocCollectionWatcher("c1", dummyWatcher);
+        latch.await(10, TimeUnit.SECONDS);
+        reader.removeDocCollectionWatcher("c1", dummyWatcher);
+
+        boolean refLazilyLoaded = false;
+        for (int j = 0; j < 10; j++) {
+          if (reader.getClusterState().getCollectionRef("c1").isLazilyLoaded()) {
+            refLazilyLoaded = true; // it should eventually be lazily loaded
+            break;
+          }
+          int attempt = j + 1;
+          log.info("ref is not lazily loaded yet. Attempt : {}", attempt);
+
+          TimeUnit.MILLISECONDS.sleep(100);
+        }
```

Comment:
TimeOut.waitFor

Reply:
ah didn't know that! Much cleaner with it 👍🏼